lib/ddtrace/transport/traces.rb

```diff
@@ -8,25 +9,173 @@ module Traces
+          super(data)
+          @trace_count = trace_count
         end
 
-        def encode_with(encoder)
-          encoder.encode_traces(data)
+        def count
+          data.length
         end
       end
 
       # Traces request
       class Request < Transport::Request
-        def initialize(traces)
-          super(Parcel.new(traces))
+        def initialize(data, trace_count)
+          super(Parcel.new(data, trace_count))
         end
       end
 
       # Traces response
       module Response
-        attr_reader :service_rates
+        attr_reader :service_rates, :trace_count
+      end
+
+      # Traces chunker
+      class Chunker
+        # Trace agent limit payload size of 10 MiB (since agent v5.11.0):
+        # https://github.com/DataDog/datadog-agent/blob/6.14.1/pkg/trace/api/api.go#L46
+        #
+        # We set the value to a conservative 5 MiB, in case network speed is slow.
+        DEFAULT_MAX_PAYLOAD_SIZE = 5 * 1024 * 1024
+
+        attr_reader :encoder, :max_size
+
+        #
+        # Single traces larger than +max_size+ will be discarded.
+        #
+        # @param encoder [Datadog::Encoding::Encoder]
+        # @param max_size [String] maximum acceptable payload size
+        def initialize(encoder, max_size: DEFAULT_MAX_PAYLOAD_SIZE)
+          @encoder = encoder
+          @max_size = max_size
+        end
+
+        # Encodes a list of traces in chunks.
+        # Before serializing, all traces are normalized. Trace nesting is not changed.
+        #
+        # @param traces [Array<Trace>] list of traces
+        # @return [Enumerator[Array[Bytes,Integer]]] list of encoded chunks: each containing a byte array and
+        #   number of traces
+        def encode_in_chunks(traces)
+          encoded_traces = traces.map { |t| encode_one(t) }.compact
+
+          Datadog::Chunker.chunk_by_size(encoded_traces, max_size).map do |chunk|
+            [encoder.join(chunk), chunk.size]
+          end
+        end
+
+        private
+
+        def encode_one(trace)
+          encoded = Encoder.encode_trace(encoder, trace)
+
+          if encoded.size > max_size
+            # This single trace is too large, we can't flush it
+            Datadog.logger.debug { "Dropping trace. Payload too large: '#{trace.map(&:to_hash)}'" }
+            # TODO: Datadog::Debug::HealthMetrics.increment('tracer.encoder.trace.encode')
+
+            return nil
+          end
+
+          encoded
+        end
+      end
+
+      # Encodes traces using {Datadog::Encoding::Encoder} instances.
+      module Encoder
+        module_function
+
+        def encode_trace(encoder, trace)
+          encoder.encode(trace.map(&:to_hash))
+        end
+      end
+
+      # Sends traces based on transport API configuration.
+      #
+      # This class initializes the HTTP client, breaks down large
+      # batches of traces into smaller chunks and handles
+      # API version downgrade handshake.
+      class Transport
```

Comment:
I move the transport to this `traces.rb` file, as it currently is tracer-specific. This should allow for a more clear separation from other tracer-agnostic parts.

Reply:
I think this is okay for now, but conceivably, other data types might want to re-use API behavior in the transport somewhere down the road. Still I like the simplification of the transport itself, and how we're separating concerns, so I'm happy with this change right now.